feat(build, syncthing): dynamic console allocation for Windows

Shablone opened the pull request:
# Dynamic Console Allocation for Windows

### Purpose
 Windows GUI builds either always show a short console pupup on windows systemstart implementation or never show one (CLI users can't see output).
See [PR10334](https://github.com/syncthing/syncthing/pull/10334) , [PR926](https://github.com/syncthing/docs/pull/926) and many more

## Solution
Implement dynamic console allocation:
- No console when double-clicked from Explorer  
- Console appears when launched with CLI arguments
- `--no-console` flag to explicitly disable console
- Console behavior preserved across monitor restarts

### Testing
```bash
syncthing.exe              # No console (GUI)
syncthing.exe --help       # Shows console + help
syncthing.exe --no-console # No console even with args
```
Opening from explorer won't show a console.


## Implementation
- Uses Windows API (`AllocConsole`/`AttachConsole`) to create console on demand
- Monitor process passes `--no-console` to child processes when appropriate
- Windows-only using build constraints

Fixes console window appearing briefly when launched from Explorer while maintaining proper CLI functionality.

## Authorship

* Name: Elias @shablone
* Email: 1elias.bauer@gmail.com

Commit:
feat: use a gui build by default and dynamically attach to console

Signed-off-by: Elias <1elias.bauer@gmail.com>

rafaeloledo commented:
Can you add my code with coauthor msg in the commit description?

```
Co-authored-by: rafaeloledo <rafaeloliveiraledo@gmail.com>
```

`build.go`
```go
if goos == "windows" {
    b.WriteString(" -H=windowsgui")
}
```

It tooks my last 3 days investigating the problem 😆

rasa commented on cmd/syncthing/monitor.go:
```diff
@@ -88,6 +88,33 @@ func (c *serveCmd) monitorMain() {
+	childArgs := args[1:] // Start with original args (excluding binary name)
+	if noConsole {
+		// Check if --no-console is already in the args
+		hasNoConsole := false
```
hasNoConsole := slices.Contains(childArgs, "--no-console")

rasa commented on cmd/syncthing/console_windows.go:
```diff
@@ -0,0 +1,115 @@
+	procFreeConsole      = kernel32.NewProc("FreeConsole")
+	procGetConsoleWindow = kernel32.NewProc("GetConsoleWindow")
+	procSetStdHandle     = kernel32.NewProc("SetStdHandle")
+	procCreateFileW      = kernel32.NewProc("CreateFileW")
```
CreateFile and SetStdHandle are in sys/x/windows

rasa commented:
One issue with this is if a user runs this GUI version from inside an SSH session, they won't see any output. The console window will appear on the desktop.

calmh reviewed (commented):
This is the approach that we didn't get to work previously. If it does work, i.e. behaves reasonably in a normal terminal etc, that's cool. I don't see why you're reinventing command line parsing multiple times instead of using the parsed command line though?

calmh commented on build.go:
```diff
@@ -901,6 +901,9 @@ func ldflags(tags []string) string {
 	b := new(strings.Builder)
 	b.WriteString("-w")
 	b.WriteString(" -buildid=")
+	if goos == "windows" {
```
How come?

rasa replied:
It doesn't. I misread.

Shablone commented:
Thank you for the review :)
I will rework that changes accordingly.

I was reinventing the stuff because
- It was late
- I'm not familiar with Go 
- I didn't get familiar with the remaining codebase
Sorry for that

I will also investigate the ssh session. Although it seems very niche to me, to ssh into a windows server.

calmh commented:
I think the SSH bit is indeed extremely niche, but probably follows the same scenario as just running the command in an existing cmd? At least the latter should work, imho.

Commit:
fix: build param order

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
fix: add copyright

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: use slices for argument handling

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: replace syscall calls with windows package functions for console handling

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
fix: improve console initialization error handling and cleanup

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: enhance openURL function to hide command window during execution

Signed-off-by: Elias <1elias.bauer@gmail.com>

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
fix: missing = in build argument

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: improve console initialization

- Improve error handling with proper Windows API error checking
- Add handle validation to prevent invalid handle usage
- Enhance FreeConsole with better state management
- Use more robust file handle creation with proper error reporting

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
fix: skip console allocation in SSH sessions

Signed-off-by: Elias <1elias.bauer@gmail.com>

Shablone commented:
I noticed that syncthing not only was starting in console mode, it also opened the browser at runtime via calling cmd.exe resulting in  another flash if no console was present yet.
I included the fix [here](https://github.com/syncthing/syncthing/pull/10335/commits/babc1053f6750324b909b9ce09ca01ff71a7ff9d) 
I haven't checked, but I guess this was the reason for your mentioned scenario @calmh

Commit:
refactor: enhance console initialization to support internal no-console flag

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: return to original argument handling

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: revert 31745352, rely monitor console on
InternalInnerProcess instead of additional checks

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
chore: clarify comments

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: revert unnecessary blank line

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: enhance Windows URL opening logic

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
fix: use console when no args if parent exists

Signed-off-by: Elias <1elias.bauer@gmail.com>

tomasz1986 commented:
I'm trying to test the PR right now, but for me, even when launched from Explorer, the console window just stays open.

This is under Windows 10 x64, and the default shell is set to Command Prompt.

calmh reviewed (commented):
Some structural notes

Commit:
fix: simplify consoleattachment logic

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: update InitConsole to return cleanup function for Windows

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: remove cleanup

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: split console attachment and allocation

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: improve comments

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: remove console initialization from helpHandler

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: replace NoConsole2 with Serve.HideConsole

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
fix: remove ssh check

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
Merge branch 'upstream-main'

Commit:
refactor: replace old hide with new dyncamic allowcation

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
Merge branch 'dynamic-console-allocation'

Shablone commented:
Alright. Please have another look

I now splitted the console initialization into attachment (Already on present) and allocation (a new one needed). 
Relying on Kong would otherwise be very complex. The console handle should already be there  if --help was passed or some invalid arguments. Unfortunately both don't reach Kongs AfterApply(). I also can't use the other hooks of kong, as they wouldn't have parsed the arguments yet.
But I like the new approach better then the manual flag check. So now I always try to to attach to a console if present. Then we can parse the arguments and only create a new console if needed.

I also tested the ssh stuff. With the current commit when I ssh from unix into windows and call synthing.exe, it will pipe the output to my ssh shell. No new console appears 👍

I also learned that FreeConsole() is not needed, as windows this stuff on its own on process termination. So I could delete the defering stuff

I removed the old HideConsole and replaced it with the dynamic allowcation

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
refactor: simplify openURL function by removing fallback to cmd.exe

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
refactor: simplify openURL function

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
Merge branch 'dynamic-console-allocation'

calmh committed:
Merge branch 'main' into main

calmh reviewed (commented):
Thanks, sorry for taking some time on this, I do want to get it in. Can you sort out the various style/linter complaints please, and then I think this is ready given it passes some manual testing 🙏

Commit:
chore: lint

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
Merge branch 'dynamic-console-allocation'

Commit:
fix: simplify consoleAttached

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
Merge branch 'dynamic-console-allocation'

Commit:
chore: fix lint

Signed-off-by: Elias <1elias.bauer@gmail.com>

Commit:
Merge branch 'dynamic-console-allocation'

calmh commented:
Actual Windows users who want this, can we have some testing? 🙏

Commit:
Merge remote-tracking branch 'upstream/main'

calmh commented:
I want this merged, but I want people who run Windows and want this to actually test it.

dotacow commented on cmd/syncthing/main.go:
```diff
@@ -314,6 +324,27 @@ func (c *serveCmd) Run() error {
+		return false
+	}
+
+	// No command line arguments without parent (Main should have called AttachConsole already)
+	// means binary was probably double-clicked -> don't allocate console
+	if len(os.Args) <= 1 {
+		return false
+	}
+
```
This method works for detecting launching directly from the primary .exe/search bar.
But for users who use shortcuts with target flags, there is more than one argument. I stumbled upon testing while trying to separate the PR's config version from my installation, I had to pass the ``-home`` flag. example of the use case:
`C:\Users\dotacow\syncthing-test\bin\syncthing.exe --home="C:\Users\dotacow\syncthing-test\test_env" "--no-console"`
it is quite a niche use case (modifying target properties through a shortcut), and anyone who wants to modify target flags, should be able to add `--no-console` as they wish, and IMO it is better to have this be the default behavior, but I have to point it out anyways.

dotacow reviewed (commented):
Tested on Windows 10; LGTM.

built using `go run build -no-upgrade`.

CLI behavior is as expected.

double clicking on the primary executable functions as expected (no console).

double clicking on a shortcut:
double clicking with property modifications shows the console. (as explained in the other comment)
double clicking with the `--no-console` flag added to properties starts the GUI and no console, no console flash for me in any double-clicked case.

calmh committed:
Merge branch 'main' into main

dotacow commented:
**Issue (potentially blocking):** when using the CLI, pressing Enter is required to get a clean prompt afterward.

example:
```powershell
PS C:\Users\dotac\syncthing> .\bin\syncthing.exe --version
PS C:\Users\dotac\syncthing> syncthing v2.0.12-dev.40.g8a267223-pr10335 "Hafnium Hornet" (go1.26.2 windows-amd64) DESKTOP-6NU0CQ9\dotac@DESKTOP-6NU0CQ9 2025-10-27 14:57:37 UTC [modernc-sqlite, noupgrade]
# <--- (cursor blinks here, and needs manual flushing)
```
and when you try typing the next command:
```powershell
PS C:\Users\dotac\syncthing> .\bin\syncthing.exe --version
PS C:\Users\dotac\syncthing> [this is me typing the next command]0.g8a267223-pr10335 "Hafnium Hornet" (go1.26.2 windows-amd64) DESKTOP-6NU0CQ9\dotac@DESKTOP-6NU0CQ9 2025-10-27 14:57:37 UTC [modernc-sqlite, noupgrade]
```

compare this to the current binary's output:
```powershell
PS C:\Users\dotac\OneDrive\Desktop\syncthing-windows-amd64-v1.28.1> .\syncthing.exe --version
syncthing v2.1.0 "Hafnium Hornet" (go1.25.9 windows-amd64) builder@github.syncthing.net 2026-05-12 05:59:47 UTC
PS C:\Users\dotac\OneDrive\Desktop\syncthing-windows-amd64-v1.28.1> # <--- cursor appears here (i.e.e normal every day terminal behaviour)
```

Hello, sorry for not catching this earlier. it wasn't really apparent through the log output, I only caught it when I recently wanted to check if I was running the latest version and accidentally ran ``--version`` with this PR's binary.

if my concerns are unfounded please let me know, been a while since I kept up with this PR, but this could potentially break piping and other automated logging systems.

I'll investigate this further.

(@calmh @Shablone  quiet ping)

Shablone commented:
Thank you for investigating!
This behavior seems familiar to me. I think I experienced it earlier in dev too. But I also thought I fixed that already :/
It really shouldn't do that. I'll take a look

Shablone commented:
Ok, this is much more troublesome than I initially thought.

Flushing the output does not really help. There is actually a bigger culprit here.

Returning the user input is not the issue. In fact, it is returned to the user even before Syncthing outputs anything.

With console attachment, Syncthing starts and is able to communicate asynchronously. Meaning the console does not wait for Syncthing to terminate. It allows new user input while still receiving output from the attached Syncthing process. For example, when using the `serve` command, it can continue outputting text over time while the user is still able to interact with the parent console — not with Syncthing itself.

Unfortunately, on Windows there are only two options. The subsystem (`CUI` vs `GUI`) is baked into the PE header at compile time. Windows either:

* `CUI`: allocates a console if not present, shell waits synchronously → **console flashes on double-click**
* `GUI`: no console, shell returns immediately. Even if we attach to the console → **async output issue**

(For our case GUI would be without actual graphics)

See the [Microsoft documentation on console allocation policy](https://learn.microsoft.com/en-us/windows/console/console-allocation-policy).

There are several forum posts about this and no real solution. There are some hacky workarounds that simulate the user prompt or kill the parent process, but I would not want to go down that path.

However, the link above also presents a suitable solution, although it is unfortunately only available on Windows 11 24H2+:
-> using a new  windows manifest. It tells Windows exactly what we want: On CUI don't open a new console if there is no parent console.

I think this PR can be closed. Dynamic console allocation would solve the GUI issue, but it would make the CUI behavior unusable.

I would open another PR solely for the change in [cmd/syncthing/openurl_windows.go](https://github.com/syncthing/syncthing/pull/10335/changes#diff-d4f0b41c9525a7b4bf68ea1629a2833e171ece58af593909bdc03cd301873b21), which is still valid for skipping `cmd` when opening the browser. However, that is actually unrelated to dynamic console allocation.

I could also open another PR for the manifest approach once I have looked into it further. That would also be unrelated to dynamic console allocation.

Still there is the possibility to ship two binaries: one CUI and one GUI.

There are also Syncthing installers, although I have not looked into them yet. They may want to check out
[some process creation flags](https://learn.microsoft.com/en-us/windows/win32/procthread/process-creation-flags)
* `CREATE_NO_WINDOW`
* `DETACHED_PROCESS`

(@calmh quiet ping)

dotacow commented:
Firstly, thanks for your work on this, it is quite a niche feature but I am nonetheless grateful to you.

but yeah your reasoning checks out to me, seems like a race condition between syncthing and the shell over the console's output buffer, but we are limited by the available Windows APIs to solve it.

regarding the multiple binary approach, I actually was going to suggest that myself, though I am sure a design choice like that requires it's own careful consideration by the maintainers.

calmh commented:
We've discussed the multiple binaries approach a few times. I even had a PR open for it for quite a while. It didn't attract a lot of positive vibes.

> I could also open another PR for the manifest approach once I have looked into it further. That would also be unrelated to dynamic console allocation.

This ^ sounds good to me.
